Hermes WebSocket API (Fixes #722)

mpforce1 opened the pull request:
# Description

Implements an integration with the new Hermes API. As per issue #722 it seems like Twitch has finally finished A/B testing the new Hermes API. This PR adds a new WebSocket based integration with this API. It acts as a drop in replacement for the older PubSub WebSocket integration.

As part of the implementation some of the handling of PubSub Messages has been abstracted. This was done to avoid duplicating the handling logic in both integrations.

I've added a new setting to the miner constructor `use_hermes`. It defaults to `False` for now to avoid changing current behaviour for most users. In the future this might be removed if Twitch drops the old PubSub API.

I've opened this PR as draft as I'm currently seeing how it acts in the long term. This is because the reconnection logic in the new integration is different and I want to make sure it works as designed. To that end, it'd be really useful if some other community members could try out this branch and see if they run into any issues. In particular, I'd like to know:

- If anyone runs into any cases where some topics don't end up getting subscribed to or they get dropped at some point later, for example during reconnection.
- If anyone finds a case where a client should reconnect but no new client is actually created. 
- If anyone finds issues when the pool manages multiple clients, for example if the pool leaves a dangling connection or more clients are created than are actually needed.
- If I've introduced a regression with the old PubSub integration. This can be tested by setting `user_hermes` to `False` or omitting it from the configuration.

I've tested all of this locally but more users testing in more varied setups should help catch any edge cases.

Fixes #722

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested on my local setup. I'm currently running a long running test to check reconnection logic. However, my test setup only connects to a few streamers.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been updated in requirements.txt

mpforce1 committed:
Fixes #722

Implements a WebSocket client and pool for the Hermes API.

Abstracts some of the handling for PubSub Messages.

Adds the "use_hermes" Setting to allow the use of the Hermes WebSocket API instead of the PubSub WebSocket API. Defaults to False.

Renamed WebSocket classes and moved them to new websocket module.

mpforce1 committed:
Updates example.py and README.md

mpforce1 commented:
> Have been using it for around 48+ hours and only encountered "websocket closed None : None error" and "websocket error 104 connection closed by peer " error otherwise all features of the code are working perfectly fine ( excluding betting and drops which are still yet to test) FYI:- I am on Linux x64 vps

Hey, thanks for your help!

The `websocket closed None : None error` message is normal. The `None`s there are the status code and reason text. `None` generally comes through there if the WebSocket gets closed locally for any reason, there should be something more useful elsewhere in the logs that'll show up as an error if there's a problem.  I might downgrade it to a `DEBUG` since it's not something you should normally have to worry about., although the original WebSocket logs it as `INFO` so we'll see.

The other message is probably fine. 104s can happen for a number of reasons but if it's reconnecting without issues I wouldn't worry about it.

G0KU0 commented:
how to download?

G0KU0 commented:
> > how to download?
> 
> You can clone my branch [here](https://github.com/mpforce1/Twitch-Channel-Points-Miner-v2/tree/hermes). No automated builds for this one since it's not an official release.

I have a question: If I upload this to rander.com after I've linked it to my Twitch account, why does it ask me to link it again if I've already linked it?

mpforce1 commented:
> I have a question: If I upload this to rander.com after I've linked it to my Twitch account, why does it ask me to link it again if I've already linked it?

Sorry, I have no idea.

mpforce1 commented:
> Update: There might be an issue somewhere related to betting cuz some of the best are getting logged twice that means maybe they are being processed twice (there might be a chance that the streamer actually made 2 different predictions back to back not sure though but still thought to drop my observations here)

Yeah, that looks like 2 different predictions with the same name happening close together. Might be worth checking the stream to see if it happens again.

mpforce1 commented:
> But the disconnections are a lot frequent as compared to pubsub like i disconnected 7-8 times in a 12 hours run and the most frequent was on a gap of like 35-40 mins only

I've seen an increase in reconnects too, not to the same degree but then I only run with a single client in the pool. I think this is something they've built into the new API to better distribute server load but no way to know for sure. As long as it reconnects without issue there should be no problem.

For that 35-40 min gap, were the disconnections from something like a timeout or were they from a `reconnect` message from Twitch? I'm asking because it could be you were just experiencing a period of poor connectivity. The latter case can be checked by looking in the `DEBUG` logs for a `reconnect` message just prior to the reconnection attempt. Otherwise, it'll be the prior case and you should see either the code and reason in the `#[client number] - WebSocket closed: [status] - [reason]` or the error in the `#[client number] - WebSocket error: [error message]` logs.

For reference, the `reconnect` message logs look like:
```
[timestamp] - DEBUG - TwitchChannelPointsMiner.classes.websocket.hermes.Client - [on_message]: #0 - Received: {"reconnect":{"url":"wss://hermes.twitch.tv/b/v1?..."},"id":"...","type":"reconnect","timestamp":"..."}
```
where the `type` is `reconnect`.

mpforce1 commented:
> @mpforce1 do you have any other social media or email address where I can share you the logs file I don't wanna share it publically here 😅

No, sorry. The last time this happened the user uploaded them to proton storage and posted a link here. Once I'd confirmed download they deleted the files. You could try something like that.

mpforce1 commented:
@Gamerns10s Got it. I'll let you know if I find anything.

mpforce1 commented:
@Gamerns10s Right, I think I might have the issue pinned down but it'll depend on some very strange stuff. Is it possible for you to inspect the `websocket-client` library in your deployed instances? I don't know how you've got it set up but, at least for me, they show up in `[base]/python3.12/site-packages/websocket` Might be somewhere like `/usr/local/lib/python3.12/site-packages/websocket`. In there, could you check the contents of `_app.py` and see if there's something in the file that matches `return closed(frame)`?

If I'm right, you've somehow got a copy of an unreleased version of the `websocket-client` library that's currently got a bug that incorrectly calls `closed` with a frame object instead of an `Exception` or `string`.

If this is all true, all you should need to to is delete that package and reinstall it, it should be version 1.8.0 released April 2024.

OlevO1 commented:
<img width="1220" height="607" alt="image" src="https://github.com/user-attachments/assets/0a0dcd47-e559-429d-a4dc-404cac1cdfc3" />
<img width="562" height="127" alt="image" src="https://github.com/user-attachments/assets/acd03ef9-0907-4d49-ba97-53c894787a78" />

OlevO1 commented:
Thanks for the help
Am I seeing it right that the new API can mine 4-5 channels at the same time?

OlevO1 commented:
Is it normal if the websocket drops every 2-3 mins?

```
13:07:41#0 - WebSocket error: Connection to remote host was lost.
13:07:41#0 - Reconnecting to Twitch Hermes server in ~30 seconds
13:08:04 Claim BladeFlex UpEmotes (BladeFlex UpEmotes) 128/120 (106%)
13:08:11 Connection to remote host was lost. - goodbye
13:08:11#0 - WebSocket closed: None - None
13:09:24+60 --> jasontheween (29.65k points) - Reason: CLAIM.
13:09:57#0 - WebSocket closed: 4100 - ping pong failed
13:09:57#0 - Reconnecting to Twitch Hermes server in ~30 seconds
13:10:38+12 --> jasontheween (29.66k points) - Reason: WATCH.
13:12:39#0 - WebSocket closed: 4100 - ping pong failed
13:12:39#0 - Reconnecting to Twitch Hermes server in ~30 seconds
```

mpforce1 commented:
> Is it normal if the websocket drops every 2-3 mins?

It shouldn't do that, but it's not something I'm seeing. It's possibility just due to poor connectivity but it might also be due to something I've just fixed.

@OlevO1 and @Gamerns10s, I just pushed a couple of changes to this branch, it might fix the `ping pong failed` issue you've both experienced. Please let me know if it keeps happening even after this change.

mpforce1 commented:
> same :(

Any chance you could post your log file for this? I'd need `file_level=logging.DEBUG` and `less=False` turn on to get the most detail.

OlevO1 commented:
here you go

[log.log](https://github.com/user-attachments/files/22699088/log.log)

OlevO1 commented:
[log2.log](https://github.com/user-attachments/files/22703533/log2.log)

Alright reinstalled everything. It’s better now. 
Is it normal that it reconnects every 40-50 mins?

mpforce1 commented:
> Is it normal that it reconnects every 40-50 mins?

Seems to be, enough people have confirmed frequent reconnections that it looks like Twitch has built it into the API. I'll do a little more investigation and see if it's something I can do anything about.

OlevO1 commented:
> > Is it normal that it reconnects every 40-50 mins?
> 
> Seems to be, enough people have confirmed frequent reconnections that it looks like Twitch has built it into the API. I'll do a little more investigation and see if it's something I can do anything about.

okay, ty

mpforce1 commented:
@OlevO1 I've pushed up a new version that should fix that dangling client issue. Let me know if it keeps happening.

If you run into that `ping pong failed` close error again, could you please run a test for me? If so, could you set `less=False` and go to `TwitchChannelPointsMiner.py` and change
```python
logging.getLogger("chardet.charsetprober").setLevel(logging.ERROR)
logging.getLogger("requests").setLevel(logging.ERROR)
logging.getLogger("werkzeug").setLevel(logging.ERROR)
logging.getLogger("irc.client").setLevel(logging.ERROR)
logging.getLogger("seleniumwire").setLevel(logging.ERROR)
logging.getLogger("websocket").setLevel(logging.ERROR)
```
to
```python
import websocket
logging.getLogger("chardet.charsetprober").setLevel(logging.ERROR)
logging.getLogger("requests").setLevel(logging.ERROR)
logging.getLogger("werkzeug").setLevel(logging.ERROR)
logging.getLogger("irc.client").setLevel(logging.ERROR)
logging.getLogger("seleniumwire").setLevel(logging.ERROR)
logging.getLogger("websocket").setLevel(logging.DEBUG)
websocket.enableTrace(True)
```
This sets the `websocket-client` log level to `DEBUG` and enables trace logging so I can see much more detail about the websocket frames. Since I'm not seeing the issue I'd really like to get a look at that to see if the library is failing to send `pong` frames for some reason.

mpforce1 commented:
> I'm using the latest version with the Hermes API implementation (PR #728), but I'm still experiencing an issue where the bot only collects points from one streamer at a time, even when multiple streamers I'm following are live simultaneously. The bot appears to connect to multiple streams based on logs, but points are only accruing for one streamer. This seems to be a limitation with how Twitch recognizes 'active viewing' rather than an issue with the Hermes API implementation itself. Is there any way to modify the script to rotate between streamers more frequently or to somehow enable simultaneous point collection from multiple streams?

It should collect from 2 streamers at once. You might be experiencing the same issue as https://github.com/rdavydov/Twitch-Channel-Points-Miner-v2/issues/722#issuecomment-3341952337 where the same channel was filling both of the miner's slots. If you want you can try to apply the fix I posted in that comment. I'm not sure how successful the fix was because I haven't heard back from the user yet.

weermis commented:
> It should collect from 2 streamers at once. You might be experiencing the same issue as [#722 (comment)](https://github.com/rdavydov/Twitch-Channel-Points-Miner-v2/issues/722#issuecomment-3341952337) where the same channel was filling both of the miner's slots. If you want you can try to apply the fix I posted in that comment. I'm not sure how successful the fix was because I haven't heard back from the user yet.

Im gonna apply that fix and run script for like 30 minutes and tell u if it worked is that okay?

weermis commented:
@mpforce1 So i ran it for a while and its working good no problem im also using my modified version of script but fused with your latest commit here and have no issues.

mpforce1 commented:
> I indeed had a unreleased version of websocket ( I didn't even knew that lol) but now I fixed it and here are my logs with the latest changes to this pr https://www.mediafire.com/file/nm0q6re1d7i58uq/Gamerns10s.20251006-014702.log/file

Hey, thanks for that. I can see you're having an issue I noticed in my own logs last night. This `Cannot open Client, wrong state: 3` error is caused by an issue with the stale connection detection. I've got a change ready that I'm gonna let run for a couple of days to see if that fixes it. I'll ping here once it's done.

mpforce1 commented:
> Just noticed that for me after the recent changes the disconnections have actually increased like in a 6 hr run I got disconnected 16 times ( 9 times in an hour and most frequent was at an interval of 4 mins only )

Yeah, that's almost certainly this stale connection detection bug. My new change a) fixes this issue b) improves resources usage by removing a `Thread` from each client and c) moves reconnection logs to `DEBUG` so end users don't get spammed with information that really isn't relevant to them.

It ran fine for me last night with no stale connections, no errors, and 2 `reconnect` requests from Twitch resulting in correct client reconnection. I still want to let it run for another day before pushing it up though, especially given the relative instability we've had on here with me pushing up every little change.

mpforce1 commented:
I've finished testing the new change, this should fix the client reconnecting all the time. The issue was twofold:
1. Twitch only sends keepalives if no other messages have been sent in a 10 second interval. However, we were only checking the keepalive message interval when looking for stale connections. Now we check the interval since the last message of any type. 
2. There was an oversight in my stale check logic where I shouldn't have combined some if clauses.

I also had a look at the Twitch web client again and tried to better match their message flow. Now the client starts `Unopened`, the pool calls `client.open` which starts the client run `Thread`. Once connection is established the client goes into an `Unwelcomed` state. Twitch should then send a `welcome` message which puts it in an `Unauthenticated` state, we also send an `authenticate` request at this point. Twitch will then respond with an `authenticateResponse` putting us in an `Open` state, we also subscribe to any pending topics at this point. At any point the connection can be considered stale if it spends too long waiting in `Unopened`, `Unwelcomed`, or `Unauthenticated` states. Previously, we were not waiting properly for the `welcome` message before authenticating. We were also running a `Thread` to do the authentication and pending subscriptions rather than responding to Twitch messages. Now we don't need to since it does those steps in response to Twitch messages.

Thanks again to everyone who's participating, you're helping us get closer to a stable implementation. Please test this new version and let me know if you have any issues.

mpforce1 commented on TwitchChannelPointsMiner/classes/websocket/hermes/Client.py:
```diff
@@ -0,0 +1,476 @@
+                    except Exception as e:
+                        # Put the failed topic back in the list and propagate the error
+                        self.pending_topics.append(topic)
+                        raise e
```
I'm actually unsure of this one. My understanding was that any exception raised from within a `with` block would automatically cause that resource to get released by calling the context manager's `__exit__` method with the exception.

mpforce1 replied:
I did some testing with a custom context manager and it looks like I'm right. The context manager's `__exit__` gets called if an arbitrary exception gets thrown from within a `with` block. In the case of a `threading.Lock` this will cause the lock to be released.

mpforce1 commented on TwitchChannelPointsMiner/utils.py:
```diff
@@ -210,3 +214,64 @@ def check_versions():
+            return _alphabet_base_36[value]
+        elif value < 62:
+            return _alphabet_base_36[value - 26].upper()
+        elif value > 62:
```
This was just a direct translation of the js client code into python. It seems to work for me, when I run it locally it generates strings that contain `_`s and `-`s.

mpforce1 commented on TwitchChannelPointsMiner/classes/websocket/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .Pool import WebSocketPool
+from .hermes import *
```
I don't think this is a problem because the `.hermes` import should only include things exported from the `hermes/__init__.py` file and that's just some specific imports from `hermes/Client.py` and `hermes/Pool.py`.

mpforce1 committed:
Updates references to python version to 3.12

mpforce1 commented:
@rdavydov I've cleaned up the commit history and updated the minimum python version in the docs/dockerfile to 3.12. I feel it's ready to be merged so I've brought it out of draft.

brunoshure commented:
What's the state of this? I'm testing this currently and doesn't seem to be working very well. Granted this is still version 2.0.3 and not the latest 2.0.4.

It's running super slow and the channels don't get processed correctly. Lots of errors. Currently I'm only able to collect RAID points since the rest doesn't work.

Gamerns10s commented:
For me it's working perfectly fine I just get a few connection reset by peer (maybe due to my server's network) sometimes and the drops aren't automatically being claimed like they are watched for the required time but aren't claimed from inventory until I restart the bot manually. Everything else is working flawlessly

mpforce1 commented:
> What's the state of this? I'm testing this currently and doesn't seem to be working very well. Granted this is still version 2.0.3 and not the latest 2.0.4.
> 
> It's running super slow and the channels don't get processed correctly. Lots of errors. Currently I'm only able to collect RAID points since the rest doesn't work.

@brunoshure 

That's strange, it's been a bit since I used this branch since I've been working on other features so I can't say if there's been some change that's made it more unstable. However, when I last left it we seemed to be in a decent state and running it quickly now I don't see any issues.

Version 2.0.4 shouldn't really make a difference to this because that change is focused on fixing something in the watch loop not the WebSocket. I actually rebased onto 2.0.4 locally a while ago but never pushed up the change, I'll just do that now.

If you're running the miner on the same setup you've been testing my other branches (thanks a lot btw 😄) it's possible there's some sort of contamination between project files or python packages that's causing the issues you're seeing. Could you please give me some more context for your issues? When you say it's running slow do you mean load time? Can you show me an example of the errors you're seeing?

brunoshure commented:
@mpforce1 

My testing repos are stored in different folders and each one has its own `venv`. Dependencies are installed and script run while inside the `venv`. So I don't think there's any contamination.

It started spitting some errors right at the start and it took a while before actually showing something in the output. After that it only showed onlines, offlines and raids. No points were actually collected. And it stayed like that for the whole run.

Errors like these:

~~~
12/12/25 00:09:12 - ERROR - [post_gql_request]: Error with GQLOperations (GetIDFromLogin): HTTPSConnectionPool(host='gql.twitch.tv', port=443): Max retries exceeded with url: /gql (Caused by NameResolutionError("HTTPSConnection(host='gql.twitch.tv', port=443): Failed to resolve 'gql.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
~~~

~~~
12/12/25 00:56:10 - ERROR - [get_spade_url]: Something went wrong during extraction of 'spade_url': HTTPSConnectionPool(host='assets.twitch.tv', port=443): Max retries exceeded with url: /config/settings.6536def2f90a2ca65bc5b05820cd3464.js (Caused by NameResolutionError("HTTPSConnection(host='assets.twitch.tv', port=443): Failed to resolve 'assets.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
~~~

~~~
12/12/25 01:06:15 - ERROR - [post_gql_request]: Error with GQLOperations (VideoPlayerStreamInfoOverlayChannel): HTTPSConnectionPool(host='gql.twitch.tv', port=443): Max retries exceeded with url: /gql (Caused by NameResolutionError("HTTPSConnection(host='gql.twitch.tv', port=443): Failed to resolve 'gql.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
~~~

~~~
12/12/25 05:16:00 - ERROR - [on_error]: #46 - 99124b62-fa38-4b14-96a5-748e72970cb9 - WebSocket error: [Errno 104] Connection reset by peer
Traceback (most recent call last):
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_app.py", line 432, in initialize_socket
    dispatcher.read(self.sock.sock, read, check)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_dispatcher.py", line 106, in read
    if not read_callback():
           ~~~~~~~~~~~~~^^
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_app.py", line 451, in read
    op_code, frame = self.sock.recv_data_frame(True)
                     ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_core.py", line 444, in recv_data_frame
    frame = self.recv_frame()
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_core.py", line 485, in recv_frame
    return self.frame_buffer.recv_frame()
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_abnf.py", line 375, in recv_frame
    self.recv_header()
    ~~~~~~~~~~~~~~~~^^
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_abnf.py", line 329, in recv_header
    header = self.recv_strict(2)
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_abnf.py", line 418, in recv_strict
    bytes_ = self.recv(min(16384, shortage))
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_core.py", line 581, in _recv
    return recv(self.sock, bufsize)
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_socket.py", line 127, in recv
    bytes_ = _recv()
  File "/home/bruno/shureandas-mpforce1-hermes/Twitch-Channel-Points-Miner-v2/venv/lib/python3.13/site-packages/websocket/_socket.py", line 99, in _recv
    return sock.recv(bufsize)
           ~~~~~~~~~^^^^^^^^^
  File "/usr/lib/python3.13/ssl.py", line 1285, in recv
    return self.read(buflen)
           ~~~~~~~~~^^^^^^^^
  File "/usr/lib/python3.13/ssl.py", line 1140, in read
    return self._sslobj.read(len)
           ~~~~~~~~~~~~~~~~~^^^^^
~~~

Logs:
[https://temp.sh/xztjO/log.zip](url)

mpforce1 commented:
@brunoshure 

> My testing repos are stored in different folders and each one has its own `venv`. Dependencies are installed and script run while inside the `venv`. So I don't think there's any contamination.

That sounds perfect 😄 

> It started spitting some errors right at the start and it took a while before actually showing something in the output. After that it only showed onlines, offlines and raids. No points were actually collected. And it stayed like that for the whole run.

It took a while to output something because it took a few hours to load all of the data for the 1000+ streamers you were testing with.

> ```
> 12/12/25 00:09:12 - ERROR - [post_gql_request]: Error with GQLOperations (GetIDFromLogin): HTTPSConnectionPool(host='gql.twitch.tv', port=443): Max retries exceeded with url: /gql (Caused by NameResolutionError("HTTPSConnection(host='gql.twitch.tv', port=443): Failed to resolve 'gql.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
> ```
> 
> ```
> 12/12/25 00:56:10 - ERROR - [get_spade_url]: Something went wrong during extraction of 'spade_url': HTTPSConnectionPool(host='assets.twitch.tv', port=443): Max retries exceeded with url: /config/settings.6536def2f90a2ca65bc5b05820cd3464.js (Caused by NameResolutionError("HTTPSConnection(host='assets.twitch.tv', port=443): Failed to resolve 'assets.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
> ```
> 
> ```
> 12/12/25 01:06:15 - ERROR - [post_gql_request]: Error with GQLOperations (VideoPlayerStreamInfoOverlayChannel): HTTPSConnectionPool(host='gql.twitch.tv', port=443): Max retries exceeded with url: /gql (Caused by NameResolutionError("HTTPSConnection(host='gql.twitch.tv', port=443): Failed to resolve 'gql.twitch.tv' ([Errno -3] Temporary failure in name resolution)"))
> ```

These errors are unrelated to changes on this branch and suggest a temporary issue with your DNS.

> ```
> 12/12/25 05:16:00 - ERROR - [on_error]: #46 - 99124b62-fa38-4b14-96a5-748e72970cb9 - WebSocket error: [Errno 104] Connection reset by peer
> ```

This sort of error `Connection reset by peer` happens when Twitch drops your connection. This could easily be caused by whatever was causing the poor DNS performance you've already seen.

I can see in your logs that all requests to `RequestBroadcastQualitiesURL` return a 404 status which is strange, I'm not sure how it even managed to collect watch streaks to be honest. Perhaps watch streaks no longer require you to actually watch the stream? At any rate, it couldn't successfully send all minute watched event requests, which is why you didn't get any points while the miner was running. To the best of my knowledge, this has nothing to do with the changes to this branch.

As a follow up, I ran the miner with around 10 streamers all yesterday with no issues.

Do you get the same issues if you run the base miner with 1000+ streamers? Do you get the same problems if you run this branch with a much smaller amount of streamers, 10s rather than 100s?

brunoshure commented:
@mpforce1 

So I ran it again with your latest force-push and everything is working. No errors whatsoever, besides the occasional `TypeError: 'NoneType' object is not subscriptable`. I didn't modify the number of streamers.

I don't know what happened but 2.0.3 was not working at all.

> For me it's working perfectly fine I just get a few connection reset by peer (maybe due to my server's network) sometimes and the drops aren't automatically being claimed like they are watched for the required time but aren't claimed from inventory until I restart the bot manually. Everything else is working flawlessly

It could be that he had made local changes but forgot about it, maybe that's why it was working fine for him. I cloned this branch fresh and did no changes. The only thing that I modified was the `run.py` file that needed the `use_hermes=True` thing.

> I'm not sure how it even managed to collect watch streaks to be honest.

It didn't. It only collected RAID points.

mpforce1 commented:
@brunoshure 

> So I ran it again with your latest force-push and everything is working. No errors whatsoever, besides the occasional `TypeError: 'NoneType' object is not subscriptable`. I didn't modify the number of streamers.
> I don't know what happened but 2.0.3 was not working at all.

My best guess is a temporary network issue but it's hard to say.

> It could be that he had made local changes but forgot about it, maybe that's why it was working fine for him. I cloned this branch fresh and did no changes. The only thing that I modified was the `run.py` file that needed the `use_hermes=True` thing.

I think they've been doing some work on a private fork, so yeah could be.

> > I'm not sure how it even managed to collect watch streaks to be honest.
> 
> It didn't. It only collected RAID points.

Oh my bad, that makes perfect sense then. Raids use a different endpoint so that one must not have been having that 404 issue.
